### PR TITLE
Add some more hash functions

### PR DESCRIPTION
**Pull request by lgoettgens**
Some progress towards https://github.com/oscar-system/Oscar.jl/issues/2222.

I pinged multiple reviewers for the different areas:
- @thofma for number theory
- @HechtiDerLachs for schemes
- @jankoboehm for modules

It would be great if you three (or other people with knowledge in the respective areas) could have a look at this. In most cases I was able to do some genuine hashing, and not just dummy hash functions.

### CODE REVIEW

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:145]: We should probably not do that. This involves choices of representatives.

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:11]:
```suggestion
```
Frankly, I do not remember from the top of my head what `domain` of `R` is in the first place.

**Review by HechtiDerLachs — COMMENTED** [review]
Thanks for the attempt! I left comments on those parts which seem to lay in my yard.

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (lgoettgens) [reply src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:11]: I was just using the `==` function defined a few lines above, and `domain(R)` is compared there. So this should be fine, doesn't it?

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (lgoettgens) [reply src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:145]: That should be something that `hash(::typeof(f))` should handle, as the `==` method for `AffineSchemeOpenSubschemeRingElem` delegates to `==` for the restrictions in https://github.com/oscar-system/Oscar.jl/blob/a502167dd30a722be879671bfa71784360302f78/src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl#L136

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:11]: I looked it up. Yes, it's fine.

**Review by thofma — APPROVED** [review]
Number theory looks fine, thanks @lgoettgens

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (HechtiDerLachs) [reply src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:145]: The restrictions are elements of localizations of quotient rings. They do not have a good normal form which can be used for hashing and even if they did, one would probably not want to trigger the required GB computation only to get the hash. So I still think we should avoid this altogether. The hash function for these ring elements here is probably forced to stay as dummy as possible.

Comment (lgoettgens) [reply src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:145]: I think I got lead astray from the `hash(::AbsLocalizedRingElem)` method, which assumes that there exist a normal form. So I did the following changes:
- change `hash(::AbsLocalizedRingElem)` not not assume that there is a normal form
- keep this hash here as is. It will then delegate to `hash(::AbsLocalizedRingElem)` (which basically does nothing by the previous point). this should be equivalent to what you want here, but more closely follows the concepts of encapsulation and delegation per dispatch.

**Comment by jankoboehm** [comment]
Thanks, fine from my side.

**Review by jankoboehm — APPROVED** [review]

**Comment by fingolfin** [comment]
@HechtiDerLachs this is just waiting for your approval. Is this PR ok from your view point?

File: src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/AffineSchemeOpenSubscheme/Rings/Methods.jl:145]:
```suggestion
```
I'm afraid I do not fully understand the last comment. I would just remove this block here. Then I think everything is good to go.

**Review by HechtiDerLachs — CHANGES_REQUESTED** [review]
~Sorry for not keeping up with this. I'm afraid we still need to change some lines (marked above). Other than that, this is good to go from my side.~

Edit: Got it now. Should be good to go, then.

**Review by HechtiDerLachs — APPROVED** [review]

**Pull request merged**
